### ISSUE TITLE
I/O module: Should we rename I/O channels to reader and writer?

### ISSUE THREAD

**Issue by mppf** [chapel-lang/chapel#18112]
Today we have a `channel` type for use in the I/O system. An I/O channel can either be a reading channel or a writing channel.

However, as part of GSoC we are looking at adding Go-style channels to Chapel (see #17954). The immediate problem is - what should these be called, since `channel` is already used in I/O?

This issue proposes that we deprecate the term `channel` for this and instead have types `reader` and `writer`. Then we can add a generic type that matches either of these (for example, `readerOrWriter` `stream`, `iostream`, or `iochannel`) that would be supported by an interface eventually but might be directly supported in the compiler for now.

E.g. in this typical use case from the fileIO primer:
```
    // Get a binary writing channel for the start of the file.
    var w = f.writer(kind=ionative);
```

it would be reasonable for the type of `w` to be `writer` rather than `channel(writing=true)`.

| Language | I/O reader equivalent | I/O writer equivalent | I/O reader or writer |
| -- | --- | --- | --- |
| C++ | istream | ostream | ios |
| Java | Reader / BufferedReader | Writer / BufferedWriter | - |
| Rust | Read / BufRead / BufReader | Write / BufWriter | - |
| Python | BufferdReader / BufferedRandom | BufferedWriter / BufferedRandom | - |
| Go | Reader / File | Writer / File | - |

C, Swift have File but not a stream idea. Julia has a similar design but seems to call the file type IOStream.

**Comment by mppf** [comment]
I've updated the issue description with a table of what other languages use. So far, I haven't found another language that uses `channel` for this purpose. `reader` and `writer` are common as is `stream` for terminology.

**Comment by mppf** [comment]
I think we should use `iostream` or `stream` for a generic type that matches `reader` and `writer`.

**Comment by bradcray** [comment]
@mppf: I think it'd be worth editing the OP to give some motivation for this change for people who find themselves here without context (as I did a few days ago).  Specifically, to free up the term `channel` for use in supporting Go-style channels as proposed in #17954.

**Comment by lydia-duncan** [comment]
Updating here: it seems like the lean internally is towards splitting into a reader and a writer type, with a reasonable subsection of the methods on channel being only relevant to one or the other

**Comment by mppf** [comment]
We had been thinking we would remove the `writing` field while introducing reader/writer types but I'm finding (at least with the initial changes) that it's more convenient in generic code to have `if ch.writing` than to have `if isSubtype(ch, writer)`. If we keep it, it can just be a parenless method that always returns true or false (for writer/reader types, respectively).

**Comment by bradcray** [comment]
> One issue that I ran into is that there are file.reader and file.writer methods. For code within the file type itself, these conflict with the type name reader / writer so I had to add helper functions with different names for use within the file type itself.

Is an implication here that user codes that may be using those routines will not receive a deprecation warning for a release before having the behavior change on them?  Or are we able to keep the old names for the purposes of deprecation and disambiguate between the two contexts within our modules for the time being?

**Comment by mppf** [comment]
>  Or are we able to keep the old names for the purposes of deprecation and disambiguate between the two contexts within our modules for the time being?

We can keep `file.reader` and `file.writer` methods but AFAIK we weren't planning to deprecate these. We can disambiguate within our own IO module for the file methods, yes.

> Is an implication here that user codes that may be using those routines will not receive a deprecation warning for a release before having the behavior change on them?

I'm not sure what you meant here since I don't see why we would want a deprecation warning for `myfile.reader()` e.g.

**Comment by bradcray** [comment]
> AFAIK we weren't planning to deprecate these.

Sorry, I misunderstood then.  I thought you were saying that since there is a conflict, we will need to change something to resolve the conflict, and imagined that we'd deprecate the methods rather than abandoning the plan to rename the types (and replace the method names with something like `getReader()` or `createReader()` or `newReader()`).

Is the actual proposal just to accept that we have both types and methods that share the same identifier?  (so maybe it's not a "conflict" so much as a name reuse or potential for ambiguity?)

**Comment by mppf** [comment]
> Is the actual proposal just to accept that we have both types and methods that share the same identifier? (so maybe it's not a "conflict" so much as a name reuse or potential for ambiguity?)

So, in slightly more detail, writing this:


``` chapel
proc file.reader(): reader(locking=false, ...) { return new reader(...); }
```

Would cause compilation errors (because the compiler will think that `reader` in this signature and body is refering to the method being defined rather than the type).

Since the conflict only matters for methods on `file` (which a user generally won't be creating), we can simply use a helper function, e.g.

``` chapel
proc file.reader(): getReaderType() { return makeNewReader(...); }
private proc getReaderType() type { return reader(locking=false, ...); }
```

Like I said, the only problem with this workaround that I know of is that the docs will look less elegant. I don't think it rises to the level of a problem that justifies renaming `file.reader` and `file.writer` methods.

**Comment by mppf** [comment]
I think another way to solve it, that might be nicer, would be to use qualified naming:

``` chapel
proc file.reader(): IO.reader(locking=false, ...) { return new IO.reader(...); }
```

**Comment by lydia-duncan** [comment]
> I don't think it rises to the level of a problem that justifies renaming `file.reader` and `file.writer` methods.

Michael pointed out to me in the IO subteam chat that we hadn't fully looked at these methods (and hadn't put them in the buckets to examine).  I suspect that we'll at least bring up the possibility of renaming them when presenting them to the team, even if the team as a whole ends up feeling the same way as Michael

**Comment by bradcray** [comment]
I woke up this morning with my subconscious continuing to chew on this.  While I understand that we can use disambiguation to distinguish these cases, it still seems unfortunate to me that we'd rely on the same identifier for these things and require disambiguation in the case that an advanced user wanted to refer to either (say, writing a tertiary method on `file`?)

Thinking about alternatives, one would be to change the methods to something like `getReader()`/`getWriter()` or `createReader()`/`createWriter()` or `newReader()`/`newWriter()`.  The other would be to change the types to something like `readerChannel`/`writerChannel` or `readChannel`/`writeChannel` or `inChannel`/`outChannel`.  I'm not crazy about any of these, but I'm not sure I'm ready to concede that we should use the same identifier for both either.

**Comment by mppf** [comment]
So, in the implementation effort, at first I tried making `record reader` and `record writer`. But I ran into challenges with that approach with getting a reasonable deprecation error for code that uses `channel` e.g. `proc MyType.readWriteThis(ch: channel)`. So, I switched implementation strategies. Now my strategy is to rename `record channel` to `record _channel` (which is no doc'd). Then we have `deprecated type channel = _channel`, all of the internal modules can still use `_channel`, and we have `type reader = _channel(writing=false, ?)` and `type writer = _channel(writing=true, ?)`.

Now why am I bringing this up? Suppose that instead of `_channel` we used a user-facing name. In [earlier discussion](https://github.com/chapel-lang/chapel/issues/18112#issuecomment-893489545), we were leaning towards `iostream`. Then, somebody making a method on a `file` can still refer to the `iostream` type as an alternative to `channel`. Since `reader` and `writer` are just type aliases like above, they can use `iostream(writing=false)` etc. without any loss of generality, and then it seems clearer (IMO) that the name collision between `file.writer` and the type `writer` is only a minor inconvenience even for somebody making a tertiary method.

I think it comes down to the question of whether or not we want reader/writer to be completely separate types or if they can be type aliases like the above. If the type alias approach is OK, we have to finalize the spelling for `iostream`. Maybe it is better spelled `ioStream` according to our conventions ?

**Comment by mppf** [comment]
> I woke up this morning with my subconscious continuing to chew on this. While I understand that we can use disambiguation to distinguish these cases, it still seems unfortunate to me that we'd rely on the same identifier for these things and require disambiguation in the case that an advanced user wanted to refer to either (say, writing a tertiary method on `file`?)
> 
> Thinking about alternatives, one would be to change the methods to something like `getReader()`/`getWriter()` or `createReader()`/`createWriter()` or `newReader()`/`newWriter()`. The other would be to change the types to something like `readerChannel`/`writerChannel` or `readChannel`/`writeChannel` or `inChannel`/`outChannel`. I'm not crazy about any of these, but I'm not sure I'm ready to concede that we should use the same identifier for both either.

I have a relatively strong negative reaction to this proposal of renaming `file.reader()` and `file.writer`. I'll try to summarize my viewpoint:
 * I like the names `file.reader()` and `file.writer()` for the methods
 * I think these methods are used more commonly than the type name `channel` or `reader`/`writer` types in the future because of type inference
 * I think it will be extremely rare for people to write tertiary methods on `file`. In contrast, I think it will be quite common to call `file.reader()` and `file.writer()` methods. So, I think we should focus on making the common use case have good names.

In contrast, I don't care deeply about the name of the type we choose as a replacement for `channel`. However I don't think that name collision between `reader` and `file.reader` is a good reason not to go with `reader`. I think other than us, it will be extremely rare to write a tertiary method on `file`, and it would be better to optimize the experience of the more common case (using the type names) than the super uncommon one (tertiary methods on `file`).

Anyway, in terms of alternative names to the type, I suppose if we continue to have `ioStream` being a type that matches either a reader or writer, we could call them `ioReader` and `ioWriter`. But personally, it'd be my preference to have `reader` and `writer` types. I suspect that after we deprecate the name `channel`, we will be able to avoid having a concept like `ioStream` at all.

**Comment by bradcray** [comment]
> I have a relatively strong negative reaction to this proposal of renaming file.reader() and file.writer()
> ...
> In contrast, I don't care deeply about the name of the type we choose as a replacement for channel
> ...

I didn't say, but I also have a preference of renaming the types over the methods if we rename either.

**Comment by mppf** [comment]
We discussed this topic a bit off issue. First, we tried to identify the concerns and to rank them. Here are the concerns we had, in order:

 * it would be nice not to create churn for user code
 * it would be nice to have method names that are verbs
 * it would be nice not to have the same type name & method name; for terminology purposes and to make it easier to define tertiary methods on file

It sounded like there was preference for not creating churn. In particular, it's very common for code doing I/O to use `.reader` or `.writer` and while these method names could perhaps be better it sounded like people would prefer not to change them, since a change wouldn't make them so much better as to justify the churn.

We did have one advocate for `ioReader` vs `reader` because it makes it clearer what the type is about. So, I will put a mini-poll here for choosing between `reader` and `writer` for the type names vs `ioReader` and `ioWriter`.

**Comment by mppf** [comment]
👍 this comment if you prefer `reader` and `writer` as the type names (which will be the same as the method names `file.reader()` and `file.writer()`).

**Comment by mppf** [comment]
👍 this comment if you prefer `ioReader` and `ioWriter` as the type names

**Comment by mppf** [comment]
👍 this comment if you want to propose something else

**Comment by mppf** [comment]
👍 this comment if you prefer `fileReader` and `fileWriter` as the type names

**Comment by bradcray** [comment]
FWIW, I swapped my vote from ioReader/Writer (which I was a pretty big fan of) to `fileReader/Writer` because it seems strictly more descriptive and accurate w.r.t. what it is and only adds two more characters.  It also avoids the modest `IO.ioReader/.ioWriter` that someone (Engin?  Vass?) pointed out in the meeting today.  That said, I'm still very open to `ioReader`/`ioWriter` as type names as well.

**Comment by e-kayrakli** [comment]
`fileReader`/`fileWriter` is the best option so far. 

The `IO.ioReader` tautology was brought up by Lydia, IIRC. My argument against `ioReader`/`ioWriter` was that we don't have `ioFile` but just `file`. Considering Michael's argument that readers and writers are just views into files, "file reader" (and "writer") is conversationally similar to "array slice".

**Comment by mppf** [comment]
@dlongnecke-cray @arezaii @vasslitvinov @benharsh @daviditen - the voting is quite close between `ioReader` and `fileReader`, so I think we will need a bit more agreement before we can proceed. We recently heard some arguments for why `fileReader` is a good name. Do any of you have an argument for why `ioReader` would be better? Thanks.

**Comment by vasslitvinov** [comment]
I do not have a preference between these two. Brad's argument above makes sense in favor or `fileReader`.

**Comment by dlongnecke-cray** [comment]
I am happy with `fileReader`.

**Comment by daviditen** [comment]
I'm also ok with `fileReader`.

**Comment by lydia-duncan** [comment]
It looks like we have quorum on `fileReader` and `fileWriter`.  Thanks, everyone!
